InformationScripting/src/queries/CompositeQuery.cpp

```diff
@@ -0,0 +1,202 @@
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "CompositeQuery.h"
+
+#include "../graph/Graph.h"
+
+namespace InformationScripting {
+
+CompositeQuery::~CompositeQuery()
+{
+	SAFE_DELETE(inNode_);
+	for (auto node : nodes_) SAFE_DELETE(node);
+	// Don't delete the outNode_ as that we delete all Graphs which we returned in the execute methods.
+}
+
+QList<Graph*> CompositeQuery::execute(QList<Graph*> input)
+{
+	inNode_->calculatedOutputs_ = input;
+	// Nodes for which we have all dependencies calculated:
+	QQueue<QueryNode*> workingSetNodes;
+	workingSetNodes.enqueue(inNode_);
+	QSet<QueryNode*> visitedNodes;
+
+	// Check if we have nodes which take no input:
+	for (auto queryNode : nodes_)
+	{
+		if (queryNode->inputs_.empty())
+		{
+			Q_ASSERT(queryNode->q_);
+			queryNode->execute();
+			workingSetNodes.enqueue(queryNode);
+		}
+	}
+
+	while (!workingSetNodes.empty())
+	{
+		auto currentNode = workingSetNodes.dequeue();
+		// TODO: once we allow cyclic dependencies we have to be smarter here:
+		if (visitedNodes.contains(currentNode)) continue;
+		visitedNodes.insert(currentNode);
+
+		for (auto outputMapping : currentNode->outputs_)
+		{
+			int outIndex = outputMapping.outputIndex_;
+			// TODO: this shouldn't be an assertion later, but rather some error which is presented to the user.
+			Q_ASSERT(outIndex < currentNode->calculatedOutputs_.size());
+
+			// FIXME: multiple output receivers requires Graph copy constructor:
+			Q_ASSERT(outputMapping.outputReceivers_.size() <= 1);
+
+			if (outputMapping.outputReceivers_.size() == 1)
+			{
+				auto receiver = *outputMapping.outputReceivers_.begin();
+				// find the input mapping of the receiver:
+				auto inputIt = std::find_if(receiver->inputs_.begin(), receiver->inputs_.end(),
+					[currentNode, outIndex] (const InputMapping& m) {
+					return m.outputFrom_ == currentNode && m.outputIndex_ == outIndex;
+				});
+				Q_ASSERT(inputIt != receiver->inputs_.end());
+				auto inputMapping = *inputIt;
+				receiver->addCalculatedInput(inputMapping.inputIndex_, currentNode->calculatedOutputs_[outIndex]);
+
+				if (receiver->canExecute())
+				{
+					receiver->execute();
+					workingSetNodes.enqueue(receiver);
+				}
+			}
+		}
+	}
+	return outNode_->calculatedOutputs_;
+}
+
+void CompositeQuery::addQuery(Query* q)
+{
+	if (!nodeForQuery(q)) nodes_.push_back(new QueryNode(q));
```

Comment:
Actually that method is never used. And I think it shouldn't be. The client should just need to say how he wants to connect the queries, the CompositeQuery should figure out on its own how it solves this. (i.e. when to create a QueryNode etc.)